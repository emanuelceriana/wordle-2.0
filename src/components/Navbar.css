.navbar {
  background-color: black;
  position: absolute;
  width: 100vw;
  height: 30px;
  top: 0;
  left: 0;
  font-size: 1.5em;
  display: flex;
  justify-content: center;
}

.navbar .title {
  position: absolute;
}

.navbar .wTitle {
  display: inline-flex;
  rotate: 180deg;
}

.navbar span:nth-child(1) {
  margin: -0.155em;
}

.navbar .actions {
  display: flex;
  flex-direction: row;
  justify-content: end;
  align-items: center;
  width: 600px;
  font-size: 0.7em;
}

.navbar .actions div {
  margin: 0.5em;
}

.navbar .actions div:hover {
  color: gold;
  cursor: pointer;
}
