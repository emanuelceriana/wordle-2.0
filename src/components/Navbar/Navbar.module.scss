.navbar {
  background-color: black;
  position: absolute;
  width: 100vw;
  height: 30px;
  top: 0;
  left: 0;
  font-size: 1.5em;
  display: flex;
  justify-content: center;

  &__title {
    position: absolute;
  }

  &__wTitle {
    display: inline-flex;
    transform: rotate(180deg);
  }

  &__actions {
    display: flex;
    flex-direction: row;
    justify-content: end;
    align-items: center;
    width: 600px;
    font-size: 0.7em;

    div {
      margin: 0.5em;

      &:hover {
        color: gold;
        cursor: pointer;
      }
    }
  }

  span {
    &:first-child {
      margin: -0.155em;
    }
  }
}

@media only screen and (max-width: 900px) {
  .navbar {
    position: sticky;
    top: 0;
    z-index: 2;
    &__wTitle {
      transform: rotate(160deg) translate(0.1em, 0.25em);
    }
  }
}
