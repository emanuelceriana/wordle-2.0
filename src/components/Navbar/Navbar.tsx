import { useCallback } from "react";
import {
  faCircleQuestion,
  faGear,
  faVolumeHigh,
  faVolumeXmark,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useContext } from "react";
import { AppContext, IAppContext } from "../../context/AppContext";
import styles from "./Navbar.module.scss";
import { UserSettings } from "../../providers/GlobalProvider";

export const Navbar = () => {
  const {
    setIsHelpMenuOpen,
    setIsSettingsMenuOpen,
    userSettings,
    setUserSettings,
  } = useContext<IAppContext>(AppContext);

  const handleFxControlClick = useCallback(() => {
    const modifiedUserSettings: UserSettings = {
      ...userSettings,
      isSoundFxActive: !userSettings.isSoundFxActive,
    };
    setUserSettings(modifiedUserSettings);
  }, [userSettings]);

  return (
    <div className={styles.navbar}>
      <div className={styles["navbar__title"]}>
        <span className={styles["navbar__wTitle"]}>〽️</span>
        <span>ordle 2.0</span>
      </div>
      <div className={styles["navbar__actions"]}>
        <div
          data-testid="openHelpMenuButton"
          onClick={() => setIsHelpMenuOpen(true)}
        >
          <FontAwesomeIcon icon={faCircleQuestion} />
        </div>
        <div
          className={styles["settingsMenu__fxControl"]}
          data-testid="soundFxControlButton"
          onClick={handleFxControlClick}
        >
          {userSettings.isSoundFxActive ? (
            <FontAwesomeIcon icon={faVolumeHigh} />
          ) : (
            <FontAwesomeIcon icon={faVolumeXmark} />
          )}
        </div>
        <div
          data-testid="openSettingsMenuButton"
          onClick={() => setIsSettingsMenuOpen(true)}
        >
          <FontAwesomeIcon icon={faGear} />
        </div>
      </div>
    </div>
  );
};

export default Navbar;
