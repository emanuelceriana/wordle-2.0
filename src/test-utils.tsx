import React, { ReactNode } from "react";
import { render } from "@testing-library/react";
import { AppContext } from "./context/AppContext";

const MockProvider: React.FC<{
  children: ReactNode | ReactNode[];
  value: object;
}> = ({ children, value }) => (
  <AppContext.Provider value={value}>{children}</AppContext.Provider>
);

export function renderWithContext(element: React.ReactElement, value: object) {
  return render(<MockProvider value={value}>{element}</MockProvider>);
}

export function withContext(element: React.ReactElement, value: object) {
  return <MockProvider value={value}>{element}</MockProvider>;
}
