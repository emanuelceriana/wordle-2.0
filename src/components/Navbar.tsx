import "./Navbar.css";
import { faGear, faCircleQuestion } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

interface NavbarProps {
  onClickHelp: () => void;
  onClickSettings: () => void;
}

export const Navbar = ({ onClickHelp, onClickSettings }: NavbarProps) => {
  return (
    <div className="navbar">
      <div className="title">
        <span className="wTitle">〽️</span>
        <span>ordle 2.0</span>
      </div>
      <div className="actions">
        <div onClick={onClickHelp}>
          <FontAwesomeIcon icon={faCircleQuestion} />
        </div>
        <div onClick={onClickSettings}>
          <FontAwesomeIcon icon={faGear} />
        </div>
      </div>
    </div>
  );
};

export default Navbar;
