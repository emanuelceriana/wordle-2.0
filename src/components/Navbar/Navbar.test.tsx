import { renderWithContext, withContext } from "../../test-utils";
import { render, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { Navbar } from "./Navbar";

interface MockValueContext {
  userSettings: {
    isSoundFxActive: boolean;
  };
  setUserSettings: jest.Mock<any, any>;
  setIsHelpMenuOpen: jest.Mock<any, any>;
  setIsSettingsMenuOpen: jest.Mock<any, any>;
}

describe("<Nabvar />", () => {
  const mockValueContext: MockValueContext = {
    userSettings: {
      isSoundFxActive: true,
    },
    setUserSettings: jest.fn(),
    setIsHelpMenuOpen: jest.fn(),
    setIsSettingsMenuOpen: jest.fn(),
  };

  const volumeHighClass = "fa-volume-high";
  const volumeMuted = "fa-volume-xmark";

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(<Navbar />, mockValueContext);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should open Help menu", () => {
    const { getByTestId } = renderWithContext(<Navbar />, mockValueContext);
    fireEvent.click(getByTestId("openHelpMenuButton"));
    expect(mockValueContext.setIsHelpMenuOpen).toHaveBeenCalled();
  });
  it("should open Settings menu", () => {
    const { getByTestId } = renderWithContext(<Navbar />, mockValueContext);
    fireEvent.click(getByTestId("openSettingsMenuButton"));
    expect(mockValueContext.setIsSettingsMenuOpen).toHaveBeenCalled();
  });
  it("should display correct icon when soundFx is Active/Deactive", () => {
    let userSettings = {
      isSoundFxActive: true,
    };

    const setUserSettings = (settings: { isSoundFxActive: boolean }) => {
      userSettings = settings;
    };

    const { getByTestId, rerender } = render(
      withContext(<Navbar />, {
        userSettings,
        setUserSettings,
      })
    );

    const soundFxIconClassesBeforeClick = getByTestId("soundFxControlButton")
      .firstElementChild?.classList;

    expect(soundFxIconClassesBeforeClick).toContain(volumeHighClass);
    fireEvent.click(getByTestId("soundFxControlButton"));

    act(() => {
      rerender(
        withContext(<Navbar />, {
          userSettings,
          setUserSettings,
        })
      );
    });

    const soundFxIconClassesAfterClick = getByTestId("soundFxControlButton")
      .firstElementChild?.classList;

    expect(soundFxIconClassesAfterClick).toContain(volumeMuted);
  });
  it("should active or deactive soundFxControl", () => {
    const { getByTestId } = renderWithContext(<Navbar />, mockValueContext);
    fireEvent.click(getByTestId("soundFxControlButton"));
    expect(mockValueContext.setUserSettings).toHaveBeenCalledWith({
      isSoundFxActive: !mockValueContext.userSettings.isSoundFxActive,
    });
  });
});
