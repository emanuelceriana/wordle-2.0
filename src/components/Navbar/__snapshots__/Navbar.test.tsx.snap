// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nabvar /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="navbar"
  >
    <div
      class="navbar__title"
    >
      <span
        class="navbar__wTitle"
      >
        〽️
      </span>
      <span>
        ordle 2.0
      </span>
    </div>
    <div
      class="navbar__actions"
    >
      <div
        data-testid="openHelpMenuButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-question "
          data-icon="circle-question"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM169.8 165.3c7.9-22.3 29.1-37.3 52.8-37.3h58.3c34.9 0 63.1 28.3 63.1 63.1c0 22.6-12.1 43.5-31.7 54.8L280 264.4c-.2 13-10.9 23.6-24 23.6c-13.3 0-24-10.7-24-24V250.5c0-8.6 4.6-16.5 12.1-20.8l44.3-25.4c4.7-2.7 7.6-7.7 7.6-13.1c0-8.4-6.8-15.1-15.1-15.1H222.6c-3.4 0-6.4 2.1-7.5 5.3l-.4 1.2c-4.4 12.5-18.2 19-30.6 14.6s-19-18.2-14.6-30.6l.4-1.2zM224 352a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="settingsMenu__fxControl"
        data-testid="soundFxControlButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-volume-high "
          data-icon="volume-high"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M533.6 32.5C598.5 85.3 640 165.8 640 256s-41.5 170.8-106.4 223.5c-10.3 8.4-25.4 6.8-33.8-3.5s-6.8-25.4 3.5-33.8C557.5 398.2 592 331.2 592 256s-34.5-142.2-88.7-186.3c-10.3-8.4-11.8-23.5-3.5-33.8s23.5-11.8 33.8-3.5zM473.1 107c43.2 35.2 70.9 88.9 70.9 149s-27.7 113.8-70.9 149c-10.3 8.4-25.4 6.8-33.8-3.5s-6.8-25.4 3.5-33.8C475.3 341.3 496 301.1 496 256s-20.7-85.3-53.2-111.8c-10.3-8.4-11.8-23.5-3.5-33.8s23.5-11.8 33.8-3.5zm-60.5 74.5C434.1 199.1 448 225.9 448 256s-13.9 56.9-35.4 74.5c-10.3 8.4-25.4 6.8-33.8-3.5s-6.8-25.4 3.5-33.8C393.1 284.4 400 271 400 256s-6.9-28.4-17.7-37.3c-10.3-8.4-11.8-23.5-3.5-33.8s23.5-11.8 33.8-3.5zM301.1 34.8C312.6 40 320 51.4 320 64V448c0 12.6-7.4 24-18.9 29.2s-25 3.1-34.4-5.3L131.8 352H64c-35.3 0-64-28.7-64-64V224c0-35.3 28.7-64 64-64h67.8L266.7 40.1c9.4-8.4 22.9-10.4 34.4-5.3z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        data-testid="openSettingsMenuButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-gear "
          data-icon="gear"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M495.9 166.6c3.2 8.7 .5 18.4-6.4 24.6l-43.3 39.4c1.1 8.3 1.7 16.8 1.7 25.4s-.6 17.1-1.7 25.4l43.3 39.4c6.9 6.2 9.6 15.9 6.4 24.6c-4.4 11.9-9.7 23.3-15.8 34.3l-4.7 8.1c-6.6 11-14 21.4-22.1 31.2c-5.9 7.2-15.7 9.6-24.5 6.8l-55.7-17.7c-13.4 10.3-28.2 18.9-44 25.4l-12.5 57.1c-2 9.1-9 16.3-18.2 17.8c-13.8 2.3-28 3.5-42.5 3.5s-28.7-1.2-42.5-3.5c-9.2-1.5-16.2-8.7-18.2-17.8l-12.5-57.1c-15.8-6.5-30.6-15.1-44-25.4L83.1 425.9c-8.8 2.8-18.6 .3-24.5-6.8c-8.1-9.8-15.5-20.2-22.1-31.2l-4.7-8.1c-6.1-11-11.4-22.4-15.8-34.3c-3.2-8.7-.5-18.4 6.4-24.6l43.3-39.4C64.6 273.1 64 264.6 64 256s.6-17.1 1.7-25.4L22.4 191.2c-6.9-6.2-9.6-15.9-6.4-24.6c4.4-11.9 9.7-23.3 15.8-34.3l4.7-8.1c6.6-11 14-21.4 22.1-31.2c5.9-7.2 15.7-9.6 24.5-6.8l55.7 17.7c13.4-10.3 28.2-18.9 44-25.4l12.5-57.1c2-9.1 9-16.3 18.2-17.8C227.3 1.2 241.5 0 256 0s28.7 1.2 42.5 3.5c9.2 1.5 16.2 8.7 18.2 17.8l12.5 57.1c15.8 6.5 30.6 15.1 44 25.4l55.7-17.7c8.8-2.8 18.6-.3 24.5 6.8c8.1 9.8 15.5 20.2 22.1 31.2l4.7 8.1c6.1 11 11.4 22.4 15.8 34.3zM256 336a80 80 0 1 0 0-160 80 80 0 1 0 0 160z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
